src/main/java/gregtech/common/items/behaviors/ItemMagnetBehavior.java

```diff
@@ -70,81 +75,98 @@ private static boolean toggleActive(ItemStack stack) {
+            World world = entity.getEntityWorld();
+            if (!drainEnergy(true, stack, energyDraw)) {
+                return;
+            }
 
-        EntityPlayer entityPlayer = (EntityPlayer) entity;
+            List<EntityItem> items = world.getEntitiesWithinAABB(EntityItem.class, new AxisAlignedBB(entity.posX, entity.posY, entity.posZ, entity.posX, entity.posY, entity.posZ).grow(range, range, range));
 
-        if (entityPlayer.isSpectator()) {
-            return;
-        }
+            boolean didMoveEntity = false;
+            for (EntityItem itemEntity : items) {
+                if (itemEntity.isDead) {
+                    continue;
+                }
 
-        List<EntityItem> itemsInRange = entityPlayer.getEntityWorld().getEntitiesWithinAABB(EntityItem.class, getAreaBoundingBox(entityPlayer));
-        for (EntityItem entityItem : itemsInRange) {
-            if (entityItem.isDead) continue;
-
-            double distanceX = (entityPlayer.getPosition().getX() + 0.5) - entityItem.posX;
-            double distanceY = (entityPlayer.getPosition().getY() + 0.5) - entityItem.posY;
-            double distanceZ = (entityPlayer.getPosition().getZ() + 0.5) - entityItem.posZ;
-            double distance = MathHelper.sqrt(distanceX * distanceX + distanceZ * distanceZ);
-            if (distance >= 0.7) {
-                if (!entityItem.cannotPickup()) {
-                    if (!drainEnergy(true, itemStack, (long) distance))
-                        return;
-
-                    drainEnergy(false, itemStack, (long) distance);
-
-                    double directionX = distanceX / distance;
-                    double directionY = distanceY / distance;
-                    double directionZ = distanceZ / distance;
-                    entityItem.motionX = directionX * speed * 8;
-                    entityItem.motionY = directionY * speed * 8;
-                    entityItem.motionZ = directionZ * speed * 8;
-                    entityItem.velocityChanged = true;
-                    entityItem.setNoPickupDelay();
+                NBTTagCompound itemTag = itemEntity.getEntityData();
+                if (itemTag != null && itemTag.hasKey("PreventRemoteMovement")) {
+                    continue;
                 }
-            } else if (!entityItem.cannotPickup()) {
-                ItemStack stack = entityItem.getItem();
-
-                ItemStack remainder = GTTransferUtils.insertItem(new ItemStackHandler(entityPlayer.inventory.mainInventory), stack, false);
-                if (remainder.isEmpty()) {
-                    entityItem.setDead();
-                } else if (stack.getCount() > remainder.getCount()) {
-                    entityItem.setItem(remainder);
+
+                if (itemEntity.getThrower() != null && itemEntity.getThrower().equals(entity.getName()) && itemEntity.pickupDelay > 0) {
```

Comment:
Should this `entity` variable be the defined `player` variable, which is the cast of `entity`?